### ISSUE TITLE
Release publishing may be significantly improved by using  Sonatype Nexus-specific APIs

### ISSUE THREAD

**Issue by pshirshov** [zio/zio#1442]
Right now we publish all our projects using native SBT functionality and `sbt-sonatype` plugin. And releases take quite a lot of time. Also release repositories may get fragmented so in some cases it may be not possible to promote a release.

I've found that using Nexus Bundles may significantly speed the process up. 

Right now my flow is:

1. I've set `publishTo` to a local directory (`publishTo := Some(Resolver.file("file", new File("target/repo")))`)
2. Once `publishSinged` finishes I have my release with proper Maven layout in my local directory
3. Then I use a simple [ANT script](https://help.sonatype.com/repomanager2/staging-releases/configuring-your-project-for-deployment#ConfiguringYourProjectforDeployment-DeploymentwiththeNexusStagingAntTasks) to publish my bundle and release the repository.

We may adopt this approach. Ideally we may implement an SBT plugin doing this (I've opened a ticket for that:  https://github.com/sbt/sbt/issues/4958),  though we may benefit a lot even from my dirty ANT script.

The script:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns:staging="antlib:org.sonatype.nexus.ant.staging">
    <property environment="env"/>
    <!-- https://oss.sonatype.org/ -->
    <property name="nexus.uri" value="${env.NEXUS_URI}" />
    <property name="nexus.user" value="${env.NEXUS_USER}" />
    <property name="nexus.pass" value="${env.NEXUS_PASS}" />
    <property name="dir.repo" value="${env.DIR_REPO}" />
    <property name="nexus.profile" value="${env.NEXUS_PROFILE}" />

    <taskdef uri="antlib:org.sonatype.nexus.ant.staging"
            resource="org/sonatype/nexus/ant/staging/antlib.xml">
    <classpath>
        <fileset dir="tasks" includes="nexus-staging-ant-tasks-*uber.jar" />
    </classpath>
    </taskdef>

    <staging:nexusStagingInfo id="target-nexus" stagingDirectory="target/staging">
        <staging:projectInfo
                groupId="org.sonatype.nexus.ant"
                artifactId="nexus-staging-ant-tasks"
                stagingProfileId="${nexus.profile}"
                version="1.0" />
        <staging:connectionInfo
            baseUrl="${nexus.uri}">
            <staging:authentication
            username="${nexus.user}"
            password="${nexus.pass}" />
        </staging:connectionInfo>
    </staging:nexusStagingInfo>

    <target name="deploy" description="Deploy: Local and Remote Staging">
        <staging:stageLocally>
            <staging:nexusStagingInfo refid="target-nexus" />
            <fileset dir="${dir.repo}" includes="**/*.*" />
        </staging:stageLocally>

        <staging:stageRemotely>
            <staging:nexusStagingInfo refid="target-nexus" />
        </staging:stageRemotely>

        <staging:releaseStagingRepository>
            <staging:nexusStagingInfo refid="target-nexus" />
        </staging:releaseStagingRepository>

        <staging:dropStagingRepository>
            <staging:nexusStagingInfo refid="target-nexus" />
        </staging:dropStagingRepository>
    </target>

</project>
``` 

The environment:
```bash
export NEXUS_URI="https://oss.sonatype.org/"
export NEXUS_USER="NEXUS-USER@gmail.com"
export NEXUS_PASS="NEXUS-PASSWORD"
# Nexus profile ID can be obtained with sbt-sonatype commands or through Nexus UI
export NEXUS_PROFILE="xxxxxxxxx"
export DIR_REPO="target/repo"
```

Short summary:

1. Instead of publishing directly to Sonatype we should publish to a local directory
2. Once we have the project published we should set the environment variables and run ant script from above
3. Project may be released in parallel - e.g. it's fine to build, upload and release 2.12 and 2.13 artifacts independently

You may find more details here: https://github.com/sbt/sbt/issues/4958 and here: https://github.com/xerial/sbt-sonatype/issues/83

You may use this project as reference: https://github.com/7mind/izumi and ask me any questions.

- [ANT dependency download & ANT invokation](https://github.com/7mind/izumi/blob/develop/.build.sh#L104)
- [ANT script](https://github.com/7mind/izumi/blob/develop/.publish.xml)
- [Azure Pipelines publish task definition](https://github.com/7mind/izumi/blob/develop/.azure-pipelines.yml#L64)
- [Any easy way to encrypt a secrets file](https://github.com/7mind/izumi/blob/develop/doc/md/flow.md#openssl-notes)
- [Secrets file extraction](https://github.com/7mind/izumi/blob/develop/.build.sh#L179)

Also there was a [plugin allowing to run ANT tasks](https://github.com/sbt/sbt-ant) directly from SBT. It's unmaintained though it may be relatively easy to port it to 1.0.

**Comment by NeQuissimus** [comment]
hmm, is this a good idea? Private APIs are usually private for a reason...

**Comment by pshirshov** [comment]
I meant "nexus-specific".

**Comment by NeQuissimus** [comment]
Aahhhh!

**Comment by pshirshov** [comment]
Seems like this issue may be eventually resolved by `sbt-sonatype`: https://github.com/xerial/sbt-sonatype/issues/89

**Comment by pshirshov** [comment]
I'll open a new ticket instead of this, sbt-sonatype can publish bundles now: https://twitter.com/taroleo/status/1169777599360331776?s=21
